### ISSUE TITLE
Usb redirection not working

### ISSUE THREAD

**Issue by Fantu** [FreeRDP/FreeRDP#3781]
I tested usb redirection with freerdp 1-2 years ago successfull except frequest crashes.
Now I retried it latest nightly build and is not working, even if I already tried to add udev rules for change usb device group to plugdev (where the user is) like what I did time ago but still give me this error:

./xfreerdp /f /bpp:32 /u:u1 /v:192.168.1.235 /dvc:urbdrc,id,dev:046d:082b
[10:41:45:582] [7721:7722] [INFO][com.freerdp.client.common.cmdline] - loading channelEx drdynvc
...
[10:41:51:631] [7721:7722] [INFO][com.freerdp.channels.drdynvc.client] - Loading Dynamic Virtual Channel urbdrc
[10:41:51:631] [7721:7722] [ERROR][com.freerdp.channels.urbdrc.client] - error processing arguments
[10:41:51:646] [7721:7722] [WARN][com.freerdp.channels.urbdrc.client] - bus:0 dev:0 not exist in udevman
[10:41:51:646] [7721:7722] [INFO][com.freerdp.channels.urbdrc.client] - VID: 0x046D, PID: 0x082B
[10:41:51:724] [7721:7722] [WARN][com.freerdp.channels.urbdrc.client] - bus:1 dev:7 not exist in udevman
[10:41:52:124] [7721:7730] [ERROR][com.freerdp.channels.urbdrc.client] - unknown FunctionId 0x1
....

the udev rules seems working:
ls -alh /dev/bus/usb/001/ total 0 drwxr-xr-x 2 root root 160 Feb 16 09:00 . drwxr-xr-x 3 root root 60 Jan 1 1970 .. crw-rw-r-- 1 root plugdev 189, 0 Feb 15 15:05 001 crw-rw-r-- 1 root plugdev 189, 1 Feb 15 15:05 002 crw-rw-r-- 1 root plugdev 189, 2 Feb 15 15:05 003 crw-rw-r-- 1 root plugdev 189, 3 Feb 15 15:05 004 crw-rw-r-- 1 root plugdev 189, 4 Feb 15 15:05 005 crw-rw-r-- 1 root plugdev 189, 6 Feb 16 09:15 007
I also tried 666 permission to devices and also run xfreerdp as sudo, and also as root, in all cases give the same error.
I also tried with debian sid freerdp2 package based on build of some months ago with same result.
I tried also on ubuntu 14.04 and on rasbian on rpi (based on jessie armhf)
I also tried to rebuild with #define HAVE_NEW_LIBUSB 1 in channel/urbdrc/client/libusb/libusb_udevice.c
I also tried /usb instead /dvc:urbdrc
I also tried with different usb device
I also tried to connect from windows rdp client to the windows 2012r2 to test if remotefx of windows side is ok and it is.
I also tried to different windows (windows 10)
I did many other tests in these 3 days without having freerdp usb redirection working.

Can someone tell me what I did wrong please?
Can someone that have it working tell me details of its enviroments and exact freerdp command line used please?

Thanks for any reply and sorry for my bad english.

**Comment by uplusplus** [comment]
try /usb:auto

**Comment by Fantu** [comment]
Thanks for reply, I already tried with /usb:auto but is not working, about devices I tried webcam and audio headset.
Any other help is appreciated.

**Comment by uplusplus** [comment]
/usb:auto: Redirect the usb device plugged in after freerdp start

**Comment by Fantu** [comment]
Thanks for reply. Retrying today with new rpi build clean I did yesterday with updated freerdp from build jessie nightly build source on rpi webcam usb redirection is working but /usb:auto not:
`sudo XAUTHORITY=/home/pi/.Xauthority DISPLAY=:0 ./xfreerdp /u:u1 /v:192.168.1.2
35 /f /bpp:32 /rfx /usb:auto /log-filters:com.freerdp.channels.urbdrc.client:DEBUG
[10:43:20:467] [3035:3036] [INFO][com.freerdp.client.common.cmdline] - loading channelEx drdynvc
Password: 
[10:43:26:815] [3035:3036] [INFO][com.freerdp.gdi] - Local framebuffer format  PIXEL_FORMAT_BGRX32
[10:43:26:815] [3035:3036] [INFO][com.freerdp.gdi] - Remote framebuffer format PIXEL_FORMAT_BGRA32
[10:43:27:137] [3035:3036] [INFO][com.freerdp.channels.drdynvc.client] - Loading Dynamic Virtual Channel urbdrc
[10:43:27:137] [3035:3036] [ERROR][com.freerdp.channels.urbdrc.client] - error processing arguments
[10:43:27:147] [3035:3036] [DEBUG][com.freerdp.channels.urbdrc.client] - UDEVMAN device registered.
[10:43:28:192] [3035:3044] [ERROR][com.freerdp.channels.urbdrc.client] - unknown FunctionId 0x1
[10:43:28:192] [3035:3044] [DEBUG][com.freerdp.channels.urbdrc.client] - ...
[10:43:28:192] [3035:3044] [DEBUG][com.freerdp.channels.urbdrc.client] - ...
[10:43:28:193] [3035:3045] [DEBUG][com.freerdp.channels.urbdrc.client] - ...
[10:43:34:859] [3035:3046] [WARN][com.freerdp.channels.urbdrc.client] - bus:1 dev:7 not exist in udevman
[10:43:34:859] [3035:3046] [DEBUG][com.freerdp.channels.urbdrc.client] - bus:1 dev:7
[10:43:34:859] [3035:3046] [ERROR][com.freerdp.channels.urbdrc.client] - libusb_device_new: ERROR!!
[10:43:54:239] [3035:3036] [INFO][com.freerdp.core] - ERRINFO_RPC_INITIATED_DISCONNECT_BY_USER (0x0000000B):The disconnection was initiated by an administrative tool on the server running in the user's session.
[10:43:54:366] [3035:3044] [INFO][com.freerdp.channels.urbdrc.client] - urbdrc_on_close: channel id 10
[10:43:54:367] [3035:3044] [DEBUG][com.freerdp.channels.urbdrc.client] - success`

Can you give me help about this problem please? Seems strange that gave libusb error also using sudo.

I have another question, is possibile redirect multiple usb devices at freerdp launch?

**Comment by uplusplus** [comment]
You can print out the error message with "libusb_strerror(errno)" to get more information.
From the source file /client/common/cmdline.c, I guess you can 'redirect multiple usb devices at freerdp launch'.You can try it.

**Comment by Fantu** [comment]
Thanks for reply, I found the "libusb_device_new: ERROR!!" at line 1944 of "channels/urbdrc/client/libusb/libusb_udevice.c" but I don't know how to add "libusb_strerror(errno)", I suppose is not valid pDev->libusb_dev because enter in a condition where is NULL and I don't know how to take the libusb error from it.
Can you or someone help me to add the error print or any other useful information please?

**Comment by morphinz** [comment]
Hello. 
Can you check dependency list for me?  

```
apt-get install uuid-dev
apt-get install libdbus-glib-1-dev
apt-get install dbus-*-dev
apt-get install libpcsclite-dev, libusb-1.0-0-dev
apt-get install libusb-1.0-0-dev
apt-get install libudev-dev
```
If you have got everything now time to compile!  
But first please remove CMakeCache.txt and write "make clean"!
  
After that you can compile it!  
cmake -DCMAKE_BUILD_TYPE=Debug -DWITH_SSE2=ON -DCHANNEL_URBDRC_CLIENT=ON -DWITH_PCSC=ON -DCHANNEL_RDPSND_CLIENT=ON  
  
This is necessary for usb redirection.. "DCHANNEL_URBDRC_CLIENT=ON"
  
Good Luck.

**Comment by Fantu** [comment]
Thanks for your reply, I rebuild it added these deps missed: libdbus-c++-dev libdbus-glib-1-dev libdconf-dbus-1-dev

And now the debug with /usb:auto is different:
```
[15:22:39:883] [8466:8467] [INFO][com.freerdp.channels.drdynvc.client] - Loading Dynamic Virtual Channel urbdrc
[15:22:39:883] [8466:8467] [ERROR][com.freerdp.channels.urbdrc.client] - error processing arguments
[15:22:39:915] [8466:8467] [DEBUG][com.freerdp.channels.urbdrc.client] - UDEVMAN device registered.
[15:22:39:334] [8466:8476] [ERROR][com.freerdp.channels.urbdrc.client] - unknown FunctionId 0x1
[15:22:39:334] [8466:8476] [DEBUG][com.freerdp.channels.urbdrc.client] - ...
[15:22:39:334] [8466:8476] [DEBUG][com.freerdp.channels.urbdrc.client] - ...
[15:22:39:334] [8466:8477] [DEBUG][com.freerdp.channels.urbdrc.client] - ...
[15:22:46:589] [8466:8478] [WARN][com.freerdp.channels.urbdrc.client] - bus:1 dev:7 not exist in udevman
[15:22:46:590] [8466:8478] [DEBUG][com.freerdp.channels.urbdrc.client] - bus:1 dev:7
[15:22:46:617] [8466:8478] [DEBUG][com.freerdp.channels.urbdrc.client] -   Port: 5
[15:22:46:623] [8466:8478] [DEBUG][com.freerdp.channels.urbdrc.client] -   DevPath: 1-1.5
[15:22:46:623] [8466:8478] [DEBUG][com.freerdp.channels.urbdrc.client] -   Hub BUS/DEV: 1 2
[15:22:46:624] [8466:8478] [DEBUG][com.freerdp.channels.urbdrc.client] - libusb_open success!
[15:22:46:624] [8466:8478] [DEBUG][com.freerdp.channels.urbdrc.client] - Regist Device: Vid: 0x045E Pid: 0x0779 InterfaceClass = 0x0E
[15:22:50:633] [8466:8476] [ERROR][com.freerdp.channels.urbdrc.client] - unknown FunctionId 0x1
[15:22:50:633] [8466:8476] [DEBUG][com.freerdp.channels.urbdrc.client] - ...
[15:22:50:633] [8466:8476] [DEBUG][com.freerdp.channels.urbdrc.client] - ...
[15:22:50:633] [8466:8493] [DEBUG][com.freerdp.channels.urbdrc.client] - ...
[15:22:55:636] [8466:8493] [DEBUG][com.freerdp.channels.urbdrc.client] - libusb_kernel_driver_active = 1
[15:22:55:716] [8466:8493] [DEBUG][com.freerdp.channels.urbdrc.client] - libusb_detach_kernel_driver = 0
[15:22:55:716] [8466:8493] [DEBUG][com.freerdp.channels.urbdrc.client] - libusb_kernel_driver_active = 0
[15:22:55:716] [8466:8493] [DEBUG][com.freerdp.channels.urbdrc.client] - libusb_kernel_driver_active = 1
[15:22:55:717] [8466:8493] [DEBUG][com.freerdp.channels.urbdrc.client] - libusb_detach_kernel_driver = 0
[15:22:55:717] [8466:8493] [DEBUG][com.freerdp.channels.urbdrc.client] - libusb_kernel_driver_active = 1
[15:22:55:717] [8466:8493] [DEBUG][com.freerdp.channels.urbdrc.client] - libusb_detach_kernel_driver = 0
```

but webcam is not visible on destination (windows)

**Comment by anandsr4566** [comment]
What is required to be done to support Mass Storage usb redirection?
We have a requirement, and we would like to help add this support.

**Comment by akallabeth** [comment]
@anandsr4566 patching the server side, mass storage is blacklisted by windows.

**Comment by anandsr4566** [comment]
Thanks for the reply. I have to look for a way, if there is to work around it.

In Windows, the USB disk is shown as a folder. 
Is this possible to do?

**Comment by AzureusNation** [comment]
Hi I'm getting this exact issue with freerdp.

CLI: xfreedrp /v:<ip address> /u:<username> /p:<password> /rfx /usb:auto

This happens with all devices i use including webcams, headsets, arduinos, and lego robots. I've run the debug filter and can see these main culprits:

[WARN][com.freerdp.channels.urbdr.client] - bus:2 dev:10 not exist in idevman
[DEBUG][com.freerdp.channels.urbdr.client] - bus:2 dev:10
[ERROR][com.freerdp.channels.urbdr.client] - libusb_device_new:  ERROR!!

These devices work fine on windows. And when running lsusb all devices appear as they should.

Please help!

**Comment by AzureusNation** [comment]
I have fixed this issue with pr #5660

**Comment by akallabeth** [comment]
Fixed with #5745

**Comment by salar1376** [comment]
hey guys can i redirect usb mass storage by editing interface class to anohter number (no equal 08)?
im saying can i use this trick to dont let windows (server) and client (linux) know this usb device ass mass storage ?
